### PR TITLE
pgserver: add more metrics for better observability

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

This PR includes doc changes to the current metrics as well as adding new metrics. With the new set of metrics, we can quantitatively analyze the read amp., write amp. and space amp. in the system, when used together with https://github.com/neondatabase/neonbench

close https://github.com/neondatabase/neon/issues/4312
ref https://github.com/neondatabase/neon/issues/4347

compaction metrics TBD, a novel idea is to print L0 file number and number of layers in the system, and we can do this in the future when we start working on compaction.

## Summary of changes

* Add `READ_NUM_FS_LAYERS` for computing read amp.
* Add `MATERIALIZED_PAGE_CACHE_HIT_UPON_REQUEST`.
* Add `GET_RECONSTRUCT_DATA_TIME`. GET_RECONSTRUCT_DATA_TIME + RECONSTRUCT_TIME + WAIT_LSN_TIME should be approximately total time of reads.
* Add `5.0` and `10.0` to `STORAGE_IO_TIME_BUCKETS` given some fsync runs slow (i.e., > 1s) in some cases.
* Some `WAL_REDO` metrics are only used when Postgres is involved in the redo process.

Example metrics:

image

The Grafana generator will automatically generate histogram visualization for a metrics. i.e., we can easily observe the p50/p90/p99/avg storage latency in the new benchmark dashboard.

image

Read amplification. With the help of newly-added metrics.

image

Space amplification.

image

Write amplification.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/metrics.rs [hunk 1]
```diff
@@ -354,6 +383,8 @@ const STORAGE_IO_TIME_BUCKETS: &[f64] = &[
     0.001000, // 1000 usec
     0.030,    // 30 ms
     1.000,    // 1000 ms
+    5.000,    // 5000 ms
+    10.000,   // 10000 ms
```
Comment (bojanserafimov) [inline pageserver/src/metrics.rs:387]: The comment "Roughly logarithmic scale" which was already not true even before your change. I think we used to have more entries but deleted some of them. Let's delete the comment.

Also see https://github.com/neondatabase/neon/issues/2848 and https://github.com/neondatabase/neon/pull/3850 @lassizci can we afford more granularity here?

Comment (bojanserafimov) [reply pageserver/src/metrics.rs:387]: As a last resort, we could make the granularity configurable, and use a finer one when doing perf tests, while keeping it as is on staging/prod

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -2417,6 +2429,7 @@ impl Timeline {
                                 Err(e) => return Err(PageReconstructError::from(e)),
                             };
                             cont_lsn = lsn_floor;
+                            *read_count += 1;
```
Comment (bojanserafimov) [inline pageserver/src/tenant/timeline.rs:2432]: Does it count as a `fs_layer` if it was cached? I'm not very familiar with the caching logic of indexes and layers but I think there is some.

**Review by bojanserafimov — APPROVED** [review]
Good stuff. We need more metrics like this.

Would it also help to measure "amount of wal processed for request" as another proxy for read amplification? Or maybe first we should do a breakdown of the read path, and see what percentage is spent on wal redo.

File: pageserver/src/metrics.rs [hunk 3]
```diff
@@ -95,6 +105,25 @@ static RECONSTRUCT_TIME: Lazy<HistogramVec> = Lazy::new(|| {
     .expect("failed to define a metric")
 });
 
+static MATERIALIZED_PAGE_CACHE_HIT_UPON_REQUEST: Lazy<IntCounterVec> = Lazy::new(|| {
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/metrics.rs:108]: Sounds good!

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:2432]: It counts for now. We can have a separate metrics to count actual I/O in one get page request (i.e., include fetching index and fetching data) later.

File: pageserver/src/metrics.rs [hunk 4]
```diff
@@ -622,7 +653,7 @@ pub static WAL_REDO_TIME: Lazy<Histogram> = Lazy::new(|| {
 pub static WAL_REDO_WAIT_TIME: Lazy<Histogram> = Lazy::new(|| {
     register_histogram!(
         "pageserver_wal_redo_wait_seconds",
-        "Time spent waiting for access to the WAL redo process",
+        "Time spent waiting for access to the WAL redo process sent to Postgres",
```
[the remark this answers is not in the document]
Comment (bojanserafimov) [reply pageserver/src/metrics.rs:656]: Doesn't "process" already imply IPC? I'm not against clarifying, I just still find the grammar confusing. Maybe "Time spent waiting for access to the Postgres WAL redo process"?

[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/metrics.rs:656]: Oh that's true 🤣 I think my brain was stuck yesterday. `Time spent waiting for access to the Postgres WAL redo process` would be a good choice.

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (bojanserafimov) [reply pageserver/src/tenant/timeline.rs:2432]: Whatever is useful for your measurements for now, let's just make it clear what we're measuring

File: pageserver/src/metrics.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/metrics.rs]: Now I only add the 30s bucket to keep the logarithmic scale.

**Comment by skyzh** [comment]
I plan to merge this today and any final comments to this PR are welcomed. Thanks for reviews!

File: pageserver/src/metrics.rs (hunk 1 above)
Comment (lassizci) [reply pageserver/src/metrics.rs]: Adding buckets to existing metrics renders the existing ones potentially quite useless, so needs to be taken in to account. Histograms, especially with timeline dimensions are very expensive, so should be used sparingly. Then if the sample size is small, they tend to be quite inaccurate too, in which case higher level summary would perhaps work better.

That being said, we can probably afford these.

**Pull request merged**